diff --git a/icu4c/source/common/Android.bp b/icu4c/source/common/Android.bp
index 6bacb5003..055be0e03 100644
--- a/icu4c/source/common/Android.bp
+++ b/icu4c/source/common/Android.bp
@@ -206,7 +206,14 @@ cc_defaults {
 
     // We deliberately do not set -DICU_DATA_DIR: ICU4C is configured on Android
     // using udata_setCommonData.
+    //
+    // We make the ICU data directory relative to $ANDROID_ROOT on Android, so both
+    // device and sim builds can use the same codepath, and it's hard to break one
+    // without noticing because the other still works.
     cflags: [
+        "-DICU_DATA_DIR_PREFIX_ENV_VAR=\"ANDROID_ROOT\"",
+        "-DICU_DATA_DIR=\"/usr/icu\"",
+
         "-D_REENTRANT",
         "-DU_COMMON_IMPLEMENTATION",
 
